r/R/dplyr-funcs.R

```diff
@@ -58,15 +58,16 @@ NULL
 #' @keywords internal
 #'
 register_binding <- function(fun_name, fun, registry = nse_funcs) {
-  name <- gsub("^.*?::", "", fun_name)
+  name <- gsub("^.*?_", "", fun_name)
   namespace <- gsub("::.*$", "", fun_name)
```

Comment:
I haven't thought about this in great detail, but I'd be expecting an implementation of this to use the `namespace` variable here, or remove it.

I'm just wondering if there might be some advantages to recording the function and namespace separately instead of appending them?  The former feels "cleaner" in my head, though that's not sufficient reason to suggest changing it.

A thought; how do we distinguish between different kinds of functions?  i.e. we have:

* bindings for functions from packages, e.g. `lubridate::as_datetime()`
* bindings directly to arrow compute functions, e.g. `arrow_utf8_split_whitespace()`
* other bindings list the one for `cast()`

In the last case here, if we have other functions which have underscores but no namespace, I think the regex would no longer be extracting the right name.

Reply:
The `namespace` variable existed there, I haven't touched it. I assume it was added for potential future use. I think it could be removed once we support the `package::` prefix. We could introduce an `arrow` specific prefix. This could be either `arrow_` or `internal_`. In which case both `arrow_fun_name` and `fun_name` would be registered.

Reply:
FWIW I think a binding name `"lubridate_date"` is better than `"date"` as it makes it clear which function we're aiming to intercept / replace in a mutate call.

Reply:
Sure - I guess what I'm envisioning involves being able to achieve that but via `registry` being more complicated than a named list, so some object which has the function's name, namespace, and the function body itself, **but** I'm not advocating for unnecessarily complicating things this way without good reason.

Reply:
Nothing that jumps out at me right now, looks good!

Reply:
Nic is absolutely right here, let's think about the design here. Especially using underscores, which are not reserved so could cause clashes | weird design decisions like they are pointing out. 

You might also look at the PR that added much of this code — there's discussion in there that is super helpful: https://github.com/apache/arrow/pull/11904#issuecomment-999258101 I'm sure you've read it as well, but there's some really good content in the jira comments as well (along with some sketches).